### ISSUE TITLE
Define default AccessType for XML configuration

### ISSUE THREAD

**Issue by dazey3** [eclipse-ee4j/jpa-api#332]
Section 2.3.1 (Default Access Type) goes into great detail to define the default access type for mapping annotations. However, it does not define a default access type strategy for XML descriptor configurations.
This causes issues for persistence providers as they are open to default to whatever strategy they want for XML configurations and overall reduces portability.

Possible fixes:

1) We pick one. Define either AccessType.FIELD or AccessType.PROPERTY as the default type for XML configuration.

2) Specifically state that `access` attribute is required in orm.xml schema.

Current orm.xml schema fragment:
```xml
    <xsd:complexType name="entity">
        <xsd:sequence>
            ...
            <xsd:attribute name="class" type="xsd:string" use="required"/>
            <xsd:attribute name="access" type="orm:access-type"/>
```
This seems pretty clean, but creates issues with backwards compatibility. If the spec can't define a default access type (because there is no annotation placement to use), this would force users to make that determination.

3) Just say that the behavior is provider dependent and address portability. This has precedent as the spec already addresses portability at several places. This would be the easiest solution as well if no one agrees we should make any changes.

**Comment by dazey3** [comment]
One other thing to note is that Section 2.3.2 does contain the following, but I don't think it is very clear what it means:

```
It is an error if a default access type cannot be determined and an access type is not explicitly specified by means of annotations or the XML descriptor.
```

Who is in error? Is this saying that # 2 is correct; that an error should be thrown by the provider for XML configurations because `'a default type cannot be determined'` and `'an access tyoe is not explicitly specified'`? Perhaps this could be cleaned up to make its point clearer

**Comment by lukasj** [comment]
2. is not an option - access in an orm.xml can be defined on multiple places - there can be a `persistence-unit-default`, a global setting or per-entity/embeddable/... setting - which one should be the one required to be set? What impact would such setting have on the overriding rules?

I think that portable behavior in this case is covered by the last paragraph of 2.3.1:

_It is an error if a default access type cannot be determined and an access type is not explicitly specified by means of annotations or the XML descriptor. The behavior of applications that mix the placement of annotations on fields and properties within an entity hierarchy without explicitly specifying the Access annotation is undefined._

which, IMHO, clearly says, that the only correct (=portable) behavior in case of missing access type strategy is to fail - access type is not explicitly specified by means of annotations or the XML descriptor => it cannot be determined.

**Comment by dazey3** [comment]
@lukasj 

> which, IMHO, clearly says, that the only correct (=portable) behavior in case of missing access type strategy is to fail - access type is not explicitly specified by means of annotations or the XML descriptor => it cannot be determined.

Interesting. I think I would agree with this interpretation. If I understand correctly, this means that if XML descriptor configuration is used, and no explicit access type is defined for an entity, then the provider "cannot determine" a default access type and should error, correct?

What I find interesting about this interpretation is that OpenJPA/EclipseLink/Hibernate all pick a default access type strategy, even in the event of XML descriptor without a strategy declaration. Perhaps the issue is that the specification is not clear enough on the expected behavior? Perhaps the specification should define the error type and be reworded?

Also, this would means the providers I listed above would need to change behavior to become compliant. I thought maybe it might be better to work with these three providers current behavior to define the spec better, but I do agree with your interpretation.

**Comment by lukasj** [comment]
Can current interpretation be: If there is no xml, then defaults defined for annotations apply and settings from XML are taken into account if and only if they are present?

**Comment by dazey3** [comment]
> If there is no xml, then defaults defined for annotations apply and settings from XML are taken into account if and only if they are present?

So, the issue isn't about mixing XML/Annotation configurations. The issue is XML only configurations (no annotations) and no explicit access declaration. What should happen in that case?

**Comment by gavinking** [comment]
@dazey I agree that the default access type for a pure-XML config is undefined, and it's very unfortunate that the leading implementations disagree on the default.

But this is hardly a disaster: pure-XML config is not so common, and if you use it, and need portability, it's really _very_ easy to stick a single `<access>FIELD</access>` or `<access>PROPERTY</access>` at the top or your `orm.xml` file.

Now, I'm not in principle against specifying that the default is `FIELD`, but that would be a breaking change for our users, and so I'm against doing it in JPA 3.2. IMO, it's a change we could consider for JPA 4.0.

For now, I don't think we need to do anything at all.

**Comment by gavinking** [comment]
> Now, I'm not in principle against specifying that the default is FIELD, but that would be a breaking change for our users, and so I'm against doing it in JPA 3.2. IMO, it's a change we could consider for JPA 4.0.


@sebersole @yrodiere this has come up a couple of times recently.

My feeling is that we agree that Hibernate has an unfortunate default here. How open are we to changing our default?

i.e. if JPA4 said that the default access type for XML mappings is `FIELD`, would that be a disaster for our users?

**Comment by yrodiere** [comment]
I suspect it would indeed be a disaster for some users who either don't read the migration guide or cannot change their mapping (people who consume JARs that contain entity/mapping produced by another team... I'm happy I'm not them).

But... major upgrades do be like that. As long as we give some global flag to revert the default to its previous value, and ideally give some heads-up (deprecation warning in an earlier version when relying on the implicit default?), I suspect they will be able to migrate?

**Comment by gavinking** [comment]
@anija-anil Could you do me a favor and confirm the behavior of EclipseLink here?

If the access type isn't explicitly specified anywhere for an entity mapped only in XML, what does EclipseLink default to? `FIELD`?

**Comment by anija-anil** [comment]
Yes. When an entity is mapped only in XML and the access type isn't explicitly specified anywhere, EclipseLink defaults to FIELD access.

**Comment by gavinking** [comment]
OK, thanks.
